### ISSUE TITLE
Kinesis: JSON deserialization bug introduced in aws-java-sdk-kinesis 1.11.160?

### ISSUE THREAD

**Issue by zackb** [aws/aws-sdk-java#1226]
When updating to release 1.11.160 of  aws-java-sdk-kinesis I started getting JSON deserialization errors in my Kinesis Stream Lambda application. It appears the class `com.amazonaws.services.kinesis.model.Record` has two setter methods for the `encryptionType` member variable. Downgrading to 1.11.159 and the issue goes away.

An error occurred during JSON parsing: java.lang.RuntimeException
java.lang.RuntimeException: An error occurred during JSON parsing
Caused by: java.io.UncheckedIOException: com.fasterxml.jackson.databind.JsonMappingException: Conflicting setter definitions for property "encryptionType": com.amazonaws.services.kinesis.model.Record#setEncryptionType(1 params) vs com.amazonaws.services.kinesis.model.Record#setEncryptionType(1 params)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Conflicting setter definitions for property "encryptionType": com.amazonaws.services.kinesis.model.Record#setEncryptionType(1 params) vs com.amazonaws.services.kinesis.model.Record#setEncryptionType(1 params)
at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:269)
at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
at com.fasterxml.jackson.databind.DeserializationContext.findContextualValueDeserializer(DeserializationContext.java:428)
at com.fasterxml.jackson.databind.deser.std.StdDeserializer.findDeserializer(StdDeserializer.java:947)
at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve(BeanDeserializerBase.java:439)
at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:296)
at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
at com.fasterxml.jackson.databind.DeserializationContext.findContextualValueDeserializer(DeserializationContext.java:428)
at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.createContextual(CollectionDeserializer.java:164)
at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.createContextual(CollectionDeserializer.java:25)
at com.fasterxml.jackson.databind.DeserializationContext.handleSecondaryContextualization(DeserializationContext.java:669)
at com.fasterxml.jackson.databind.DeserializationContext.findContextualValueDeserializer(DeserializationContext.java:430)
at com.fasterxml.jackson.databind.deser.std.StdDeserializer.findDeserializer(StdDeserializer.java:947)
at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve(BeanDeserializerBase.java:439)
at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:296)
at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:461)
at com.fasterxml.jackson.databind.ObjectReader._findRootDeserializer(ObjectReader.java:1749)
at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:1506)
at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1102)
Caused by: java.lang.IllegalArgumentException: Conflicting setter definitions for property "encryptionType": com.amazonaws.services.kinesis.model.Record#setEncryptionType(1 params) vs com.amazonaws.services.kinesis.model.Record#setEncryptionType(1 params)
at com.fasterxml.jackson.databind.introspect.POJOPropertyBuilder.getSetter(POJOPropertyBuilder.java:300)
at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.filterBeanProps(BeanDeserializerFactory.java:592)
at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.addBeanProps(BeanDeserializerFactory.java:488)
at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.buildBeanDeserializer(BeanDeserializerFactory.java:229)
at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.createBeanDeserializer(BeanDeserializerFactory.java:142)
at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2(DeserializerCache.java:403)
at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:352)
at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264)
... 22 more

**Comment by shorea** [comment]
I wouldn't be opposed to adding a @JsonIgnore annotation on the overloads of the setter. This somewhat leaks our dependency on Jackson but I can't think of another away around this issue that is backwards compatible. 

A Jackson Mixin would also work but that pushes the burden on each customer to workaround this so it doesn't seem like the right approach.

**Comment by mransley** [comment]
I have just hit this myself.  There are two getters on the bean object Record which causes lambda to fall over when using a Kinesis stream.

com.amazonaws.services.kinesis.model

> public void setEncryptionType(EncryptionType encryptionType)

and 

> public void setEncryptionType(String encryptionType)

Code appears to be triggering before it hits the lambda code itself so is a show stopper for this type of function I think.

**Comment by kijeungc** [comment]
I got exactly same issue. Any updates on this??

**Comment by shorea** [comment]
We are actively working on this and will be releasing a fix today.

**Comment by shorea** [comment]
Unfortunately the @JsonIgnore annotation did not work since Lambda configures their ObjectMapper to ignore annotations. We decided to make a breaking change and remove the overload that takes the enum to fix this issue. Going forward, we will make changes in our code generator to strictly follow the POJO standard to avoid this situation in the future. I'll get a release out today.

**Comment by shorea** [comment]
This has been fixed in 1.11.167.

**Comment by varunnvs92** [comment]
Hi All, 
As you might know we have added jackson annotations for enum getters/setters to avoid the "Conflicting setter definitions for property" error in 1.11.167. But this has broken some customers. We decided the change was backwards incompatible and it will be reverted in the next version (1.11.208). If anyone is relying on those annotations to avoid the conflicting setter error, you will be broken by this change. There are 2 options to avoid this:
1) Upgrade to Jackson 2.7 which uses heuristic rules to resolve the proper method
2) Add MixIns for all classes (that contains enum) you want to serialize or deserialize. Then configure the object mapper with those mixins.
